algorithms/kernel/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi_impl.i

```diff
@@ -144,145 +114,119 @@ Status KNNClassificationPredictKernelUCAPI<algorithmFpType>::
-            }
-            QuickSelectIndexed::select( distances, categories, rndSeq, dataBlockSize, nK, curProbeBlockSize,
-                                        curDataBlockSize, curDataBlockSize, selectResult, &st);  DAAL_CHECK_STATUS_VAR(st);
-            copyPartialSelections(context, gather_selection, selectResult.values, selectResult.indices, partialDistances,
-                                  partialCategories, curProbeBlockSize, nK, partCount, nParts, dblock != 0 && partCount == 0, &st);  DAAL_CHECK_STATUS_VAR(st);
+            computeDistances(context, curData, curProbes, distances, curDataBlockSize, curProbeBlockSize, nFeatures, &st);
+            DAAL_CHECK_STATUS_VAR(st);
+            // Select nK smallest distances and their labels from every row of the [curProbeBlockSize]x[curDataBlockSize] block
+            selector->selectLabels(distances, dataLabels, nK, curProbeBlockSize, curDataBlockSize, curDataBlockSize, 0, selectResult, &st);
+            DAAL_CHECK_STATUS_VAR(st);
+            // copy block results to buffer in order to get merged with the same selection algorithm (up to nParts of partial results)
+            // and keep the first part containing previously merged result if exists
+            copyPartialSelections(context, gather_selection, selectResult.values, selectResult.indices, partialDistances, partialCategories,
+                                  curProbeBlockSize, nK, partCount, nParts, dblock != 0 && partCount == 0, &st);
+            DAAL_CHECK_STATUS_VAR(st);
             partCount++;
-            if(partCount >= nParts)
+            if (partCount >= nParts)
             {
-                QuickSelectIndexed::select(partialDistances, partialCategories, rndSeq, dataBlockSize, nK, curProbeBlockSize,
-                                            nK * partCount, nK * nParts, selectResult, &st);  DAAL_CHECK_STATUS_VAR(st);
+                // Select nK closest neighbors from [curProbeBlockSize]x[nK * partCount] buffers with partial results
+                selector->selectLabels(partialDistances, partialCategories, nK, curProbeBlockSize, nK * partCount, nK * nParts, nK * nParts,
+                                       selectResult, &st);
+                DAAL_CHECK_STATUS_VAR(st);
                 partCount = 0;
             }
+            DAAL_CHECK_STATUS_VAR(labels->releaseBlockOfRows(labelRows));
+            DAAL_CHECK_STATUS_VAR(points->releaseBlockOfRows(dataRows));
         }
-        if(partCount > 0)
-        {
-            QuickSelectIndexed::select(partialDistances, partialCategories, rndSeq, dataBlockSize, nK, curProbeBlockSize,
-                                        nK * partCount, nK * nParts, selectResult, &st);  DAAL_CHECK_STATUS_VAR(st);
-        }
-        RadixSort::sort(selectResult.indices, sorted,  radix_buffer, curProbeBlockSize, nK, nK, &st); DAAL_CHECK_STATUS_VAR(st);
+        if (partCount > 0)
```

Comment:
I suggest to encapsulate logic of this `if` statement and `if` from line 179 in some class, and call the method of this class here. In this case, `partCount` variable will be the private state of this object.